01_oop_and_ood/src/Shape.ts

```diff
@@ -0,0 +1,63 @@
+import { Point } from "./Point";
+
+export abstract class Shape {
+    protected color: string;
+    protected filled: boolean;
+    points: Point[];
+
+    constructor(points: Point[])
+    constructor(points: Point[], color: string, filled: boolean)
+    constructor(...args: any[]) {
+        if (args.length === 1 && Array.isArray(args[0])) { 
+            this.shapeLengthCheck(args[0]);
+            this.color = 'green'; 
+            this.filled = true; 
+            this.points = args[0];
+        } else if (
+            args.length === 3 && 
+            Array.isArray(args[0]) &&
+            typeof args[1] === 'string' &&
+            typeof args[2] === 'boolean'
+            ) { 
+            this.shapeLengthCheck(args[0]);
+            this.color = args[1]; 
+            this.filled = args[2]; 
+            this.points = args[0];
+        } else { 
+            throw new Error('Invalid arguments.') 
+        } 
+    }
+
+    private shapeLengthCheck(arr) {
+        if (arr.length < 3) {
+            throw new Error('Invalid points quantity') 
+        }
+    }
+
+    toString() {
+        const filledStr = this.filled ? 'filled' : 'not filled';
+        const points = this.points.reduce((acc, curr) => {
+            return acc.concat(`(${curr.x}, ${curr.y})`)
+        }, []);
+        
+        return `A Shape with color of ${this.color} and ${filledStr}. Points: ${points.join(', ')}.`;
```

Comment:
nit: you can do
```
this.points.reduce((acc, curr) => {
            return acc.concat(`(${curr.x}, ${curr.y})`)
        }, []).join(', ')
```
and use just variable `points` at line 43

Reply:
done